File: leetcode/problems/language-specific/go/383.ransom-note.go
/*
 * @lc app=leetcode id=383 lang=golang
 *
 * [383] Ransom Note
 */

// @lc code=start
func canConstruct(ransomNote string, magazine string) bool {
	if len(ransomNote) > len(magazine) {
		return false
	}

	mappings := [26]int{}
	for _, v := range magazine {
		mappings[int(v-'a')]++
	}

	for _, v := range ransomNote {
		mappings[int(v-'a')]--
		if mappings[int(v-'a')] < 0 {
			return false
		}
	}

	return true
}

// @lc code=end

// * This is slower
func canConstructHashMapApproach(ransomNote string, magazine string) bool {
	if len(ransomNote) > len(magazine) {
		return false
	}

	mappings := map[byte]int{}
	for _, v := range magazine {
		mappings[byte(v)]++
	}

	for _, v := range ransomNote {
		mappings[byte(v)]--
		if mappings[byte(v)] < 0 {
			return false
		}
	}

	return true
}

File: leetcode/problems/language-specific/python/383.ransom-note.py
#
# @lc app=leetcode id=383 lang=python3
#
# [383] Ransom Note
#

# @lc code=start

# * Hash Table Solution | O(n) Time | O(1) Space
# * n -> The length of ransomNote

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(ransomNote) > len(magazine):
            return False

        # * Contains only 26 lowercase letters which makes it a constant space
        magazine_counter = collections.Counter(magazine)
        for c in ransomNote:
            magazine_counter[c] -= 1
            if magazine_counter[c] < 0:
                return False

        return True

# @lc code=end
